File: MyApp/PatientContext.swift
import Foundation

/// Compact, structured patient context for the "Prepare for Next Session"
/// Edge Function (Phase 2). Built from the models already in memory —
/// `Patient`, `Session`, `CompletedQuestionnaire`, and the saved
/// `WhisperService.CBTSessionAnalysis` reviews — and encoded straight to
/// JSON with snake_case keys.
///
/// Deliberately excludes identifying data (name, database IDs). Follow-up
/// questions appear only in `openFollowUps`, not inside the reviews, so
/// nothing is sent twice.
nonisolated struct PatientContext: Encodable {

    // MARK: - Sections

    /// Who the patient is, clinically. The app currently stores only the
    /// therapist's free-text patient notes; the other fields are filled in
    /// by the caller when known and omitted from the JSON otherwise.
    struct Overview: Encodable {
        var age: Int?
        var gender: String?
        var presentingProblems: String?
        var treatmentGoals: String?
        /// The therapist's free-text notes about the patient.
        var background: String?

        enum CodingKeys: String, CodingKey {
            case age
            case gender
            case presentingProblems = "presenting_problems"
            case treatmentGoals = "treatment_goals"
            case background
        }
    }

    /// One GAD-7 + PHQ-9 assessment. Answers keep their question positions
    /// (unanswered questions encode as null).
    struct Assessment: Encodable {
        let date: String
        let gad7Answers: [Int?]
        let gad7Score: Int
        let phq9Answers: [Int?]
        let phq9Score: Int
        let interferenceLevel: Int?

        enum CodingKeys: String, CodingKey {
            case date
            case gad7Answers = "gad7_answers"
            case gad7Score = "gad7_score"
            case phq9Answers = "phq9_answers"
            case phq9Score = "phq9_score"
            case interferenceLevel = "interference_level"
        }
    }

    /// A session's post-session notes.
    struct SessionNote: Encodable {
        let date: String
        let notes: String
    }

    /// The clinically relevant core of a saved AI session review. Reuses the
    /// analysis types as-is.
    struct SessionReview: Encodable {
        let date: String
        let summary: String
        let possibleNats: [WhisperService.PossibleNAT]
        let cbtCycles: [WhisperService.NextSessionPreparation.CBTCycle]
        let therapistHypotheses: [WhisperService.TherapistHypothesis]
        let assignmentsForNextWeek: [WhisperService.AssignmentForNextWeek]

        enum CodingKeys: String, CodingKey {
            case date
            case summary
            case possibleNats = "possible_nats"
            case cbtCycles = "cbt_cycles"
            case therapistHypotheses = "therapist_hypotheses"
            case assignmentsForNextWeek = "assignments_for_next_week"
        }
    }

    /// A follow-up question not yet marked discussed or not relevant.
    struct OpenFollowUp: Encodable {
        let sessionDate: String
        let question: String
        let reason: String

        enum CodingKeys: String, CodingKey {
            case sessionDate = "session_date"
            case question
            case reason
        }
    }

    /// The therapist's own formulation. Not stored in the app yet; provided
    /// by the caller when available.
    struct Formulation: Encodable {
        var formulation: String?
        var treatmentGoals: String?
        var currentFocus: String?

        enum CodingKeys: String, CodingKey {
            case formulation
            case treatmentGoals = "treatment_goals"
            case currentFocus = "current_focus"
        }
    }

    struct Metadata: Encodable {
        let sessionCount: Int
        let lastSessionDate: String?
        let upcomingFocus: String?

        enum CodingKeys: String, CodingKey {
            case sessionCount = "session_count"
            case lastSessionDate = "last_session_date"
            case upcomingFocus = "upcoming_focus"
        }
    }

    // MARK: - Payload

    let overview: Overview
    /// Last assessments, newest first.
    let assessments: [Assessment]
    /// Last sessions' notes, newest first.
    let recentSessions: [SessionNote]
    /// Last saved AI reviews, newest first.
    let recentReviews: [SessionReview]
    let openFollowUps: [OpenFollowUp]
    let formulation: Formulation
    let metadata: Metadata

    enum CodingKeys: String, CodingKey {
        case overview
        case assessments
        case recentSessions = "recent_sessions"
        case recentReviews = "recent_reviews"
        case openFollowUps = "open_follow_ups"
        case formulation
        case metadata
    }

    // MARK: - Building

    private static let maxAssessments = 6
    private static let maxSessionNotes = 5
    private static let maxReviews = 5

    /// Assembles the context from data already loaded for the patient.
    /// `questionnaires` is the patient's questionnaire history (any order),
    /// typically `store.cachedQuestionnaires(for: patient)`.
    @MainActor
    static func make(for patient: Patient,
                     questionnaires: [CompletedQuestionnaire],
                     overview: Overview? = nil,
                     formulation: Formulation = Formulation(),
                     upcomingFocus: String? = nil) -> PatientContext {

        let sessions = patient.sessions.sorted { $0.date > $1.date }

        let trimmedPatientNotes = patient.notes.trimmingCharacters(in: .whitespacesAndNewlines)
        var resolvedOverview = overview ?? Overview()
        if resolvedOverview.background == nil, !trimmedPatientNotes.isEmpty {
            resolvedOverview.background = trimmedPatientNotes
        }

        let assessments = questionnaires
            .sorted { $0.answeredDate > $1.answeredDate }
            .prefix(maxAssessments)
            .map { record in
                Assessment(
                    date: dateString(record.answeredDate),
                    gad7Answers: record.questionnaire.gad7Answers,
                    gad7Score: record.questionnaire.gad7Score,
                    phq9Answers: record.questionnaire.phq9Answers,
                    phq9Score: record.questionnaire.phq9Score,
                    interferenceLevel: record.questionnaire.interferenceLevel
                )
            }

        let recentSessions = sessions
            .filter { !$0.notes.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .prefix(maxSessionNotes)
            .map { SessionNote(date: dateString($0.date), notes: $0.notes) }

        let recentReviews = sessions
            .compactMap { session in
                session.structuredNotes.map { analysis in
                    SessionReview(
                        date: dateString(session.date),
                        summary: analysis.sessionSummary,
                        possibleNats: analysis.possibleNats,
                        cbtCycles: analysis.cbtCycles,
                        therapistHypotheses: analysis.therapistHypotheses,
                        assignmentsForNextWeek: analysis.assignmentsForNextWeek
                    )
                }
            }
            .prefix(maxReviews)

        let openFollowUps = sessions.flatMap { session in
            (session.structuredNotes?.followUpQuestions ?? [])
                .filter { $0.status != .discussed && $0.status != .notRelevant }
                .map {
                    OpenFollowUp(sessionDate: dateString(session.date),
                                 question: $0.question,
                                 reason: $0.reason)
                }
        }

        return PatientContext(
            overview: resolvedOverview,
            assessments: Array(assessments),
            recentSessions: Array(recentSessions),
            recentReviews: Array(recentReviews),
            openFollowUps: openFollowUps,
            formulation: formulation,
            metadata: Metadata(
                sessionCount: patient.sessions.count,
                lastSessionDate: sessions.first.map { dateString($0.date) },
                upcomingFocus: upcomingFocus
            )
        )
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    private static func dateString(_ date: Date) -> String {
        dateFormatter.string(from: date)
    }
}

File: android/app/src/main/java/com/cbtipul/app/data/PatientContext.kt
package com.cbtipul.app.data

import com.cbtipul.app.model.AssignmentForNextWeek
import com.cbtipul.app.model.CBTCycle
import com.cbtipul.app.model.CompletedQuestionnaire
import com.cbtipul.app.model.FollowUpStatus
import com.cbtipul.app.model.Patient
import com.cbtipul.app.model.SessionType
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@Serializable
data class PatientContext(
    val overview: Overview,
    val assessments: List<Assessment>,
    @SerialName("recent_sessions") val recentSessions: List<SessionNote>,
    @SerialName("recent_reviews") val recentReviews: List<SessionReview>,
    @SerialName("open_follow_ups") val openFollowUps: List<OpenFollowUp>,
    val formulation: Formulation = Formulation(),
    val metadata: Metadata,
) {
    @Serializable
    data class Overview(
        val age: Int? = null,
        val gender: String? = null,
        @SerialName("presenting_problems") val presentingProblems: String? = null,
        @SerialName("treatment_goals") val treatmentGoals: String? = null,
        val background: String? = null,
    )

    @Serializable
    data class Assessment(
        val date: String,
        @SerialName("gad7_answers") val gad7Answers: List<Int?>,
        @SerialName("gad7_score") val gad7Score: Int,
        @SerialName("phq9_answers") val phq9Answers: List<Int?>,
        @SerialName("phq9_score") val phq9Score: Int,
        @SerialName("interference_level") val interferenceLevel: Int? = null,
        @SerialName("gad7_notes") val gad7Notes: List<String> = emptyList(),
        @SerialName("phq9_notes") val phq9Notes: List<String> = emptyList(),
        @SerialName("interference_note") val interferenceNote: String? = null,
    )

    @Serializable
    data class SessionNote(
        val date: String,
        val notes: String,
        val type: SessionType? = null,
    )

    @Serializable
    data class SessionReview(
        val date: String,
        val summary: String,
        @SerialName("possible_nats") val possibleNats: List<com.cbtipul.app.model.PossibleNAT>,
        @SerialName("cbt_cycles") val cbtCycles: List<com.cbtipul.app.model.CBTCycle>,
        @SerialName("therapist_hypotheses") val therapistHypotheses: List<com.cbtipul.app.model.TherapistHypothesis>,
        @SerialName("assignments_for_next_week") val assignmentsForNextWeek: List<AssignmentForNextWeek>,
    )

    @Serializable
    data class OpenFollowUp(
        @SerialName("session_date") val sessionDate: String,
        val question: String,
        val reason: String,
    )

    @Serializable
    data class Formulation(
        val formulation: String? = null,
        @SerialName("treatment_goals") val treatmentGoals: String? = null,
        @SerialName("current_focus") val currentFocus: String? = null,
        @SerialName("key_automatic_thoughts") val keyAutomaticThoughts: List<String> = emptyList(),
        @SerialName("maintaining_behaviors") val maintainingBehaviors: List<String> = emptyList(),
        @SerialName("key_cbt_cycle") val keyCbtCycle: CBTCycle? = null,
    )

    @Serializable
    data class Metadata(
        @SerialName("session_count") val sessionCount: Int,
        @SerialName("last_session_date") val lastSessionDate: String? = null,
        @SerialName("upcoming_focus") val upcomingFocus: String? = null,
    )

    companion object {
        private val dateOnly = SimpleDateFormat("yyyy-MM-dd", Locale.US)

        fun make(patient: Patient, questionnaires: List<CompletedQuestionnaire>): PatientContext {
            val sessions = patient.sessions.sortedByDescending { it.date.time }
            val background = patient.notes.trim().ifEmpty { null }
            val assessments = questionnaires
                .sortedByDescending { it.answeredDate.time }
                .map { record ->
                    Assessment(
                        date = dateOnly.format(record.answeredDate),
                        gad7Answers = record.questionnaire.gad7Answers,
                        gad7Score = record.questionnaire.gad7Score,
                        phq9Answers = record.questionnaire.phq9Answers,
                        phq9Score = record.questionnaire.phq9Score,
                        interferenceLevel = record.questionnaire.interferenceLevel,
                        gad7Notes = record.questionnaire.gad7Notes,
                        phq9Notes = record.questionnaire.phq9Notes,
                        interferenceNote = record.questionnaire.interferenceNote.trim().ifEmpty { null },
                    )
                }
            val recentSessions = sessions.map {
                SessionNote(
                    date = dateOnly.format(it.date),
                    notes = it.notes,
                    type = it.type,
                )
            }
            val recentReviews = sessions.mapNotNull { session ->
                session.structuredNotes?.let { analysis ->
                    SessionReview(
                        date = dateOnly.format(session.date),
                        summary = analysis.sessionSummary,
                        possibleNats = analysis.possibleNats,
                        cbtCycles = analysis.cbtCycles,
                        therapistHypotheses = analysis.therapistHypotheses,
                        assignmentsForNextWeek = analysis.assignmentsForNextWeek,
                    )
                }
            }
            val openFollowUps = sessions.flatMap { session ->
                (session.structuredNotes?.followUpQuestions ?: emptyList())
                    .filter { it.status != FollowUpStatus.Discussed && it.status != FollowUpStatus.NotRelevant }
                    .map {
                        OpenFollowUp(dateOnly.format(session.date), it.question, it.reason)
                    }
            }
            return PatientContext(
                overview = Overview(background = background),
                assessments = assessments,
                recentSessions = recentSessions,
                recentReviews = recentReviews,
                openFollowUps = openFollowUps,
                metadata = Metadata(
                    sessionCount = patient.sessions.size,
                    lastSessionDate = sessions.firstOrNull()?.let { dateOnly.format(it.date) },
                ),
                formulation = PatientContext.Formulation(
                    formulation = patient.formulation?.therapistHypothesis?.trim()?.ifEmpty { null },
                    treatmentGoals = patient.formulation?.treatmentGoal?.trim()?.ifEmpty { null },
                    currentFocus = patient.formulation?.coreBelief?.trim()?.ifEmpty { null },
                    keyAutomaticThoughts = patient.formulation?.keyAutomaticThoughts.orEmpty()
                        .map { it.trim() }.filter { it.isNotEmpty() },
                    maintainingBehaviors = patient.formulation?.maintainingBehaviors.orEmpty()
                        .map { it.trim() }.filter { it.isNotEmpty() },
                    keyCbtCycle = patient.formulation?.keyCBTCycle,
                ),
            )
        }

        fun lastSessionAssignments(patient: Patient, now: Date = Date()): List<AssignmentForNextWeek>? {
            val startOfTomorrow = java.util.Calendar.getInstance().apply {
                time = now
                set(java.util.Calendar.HOUR_OF_DAY, 0)
                set(java.util.Calendar.MINUTE, 0)
                set(java.util.Calendar.SECOND, 0)
                set(java.util.Calendar.MILLISECOND, 0)
                add(java.util.Calendar.DAY_OF_YEAR, 1)
            }.time
            val assignments = patient.sessions
                .filter { it.date < startOfTomorrow }
                .maxByOrNull { it.date.time }
                ?.structuredNotes
                ?.assignmentsForNextWeek
            return assignments?.takeIf { it.isNotEmpty() }
        }
    }
}
